## OMAR ELALEM 8937288

In [9]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import datasets
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn import metrics
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import make_pipeline
sns.set_style("whitegrid")

In [3]:
X, y = datasets.load_diabetes(as_frame=True, return_X_y=True)

In [4]:
degrees = np.arange(0, 9) 
for degree in degrees:
  model = make_pipeline(PolynomialFeatures(degree), LinearRegression())

In [18]:
cross_val_scores = [] #creating the empty list
for degree in degrees:
    # Create a pipeline that creates polynomial features, then applies linear regression
    model = make_pipeline(PolynomialFeatures(degree), LinearRegression())
    
    # Use cross-validation to compute the R-square(R^2) across multiple splits of the data
    score = cross_val_score(model, X, y, cv=5,scoring="r2")
    rSqr_mean = np.mean(score) # Calculate the mean R-square across all folds
    rSqr_std = np.std(score) # Calculate the standar-deviation R-square across all folds
    # Use cross-validation to compute the Mean Ablsoute Error(MAE) across multiple splits of the data
    scores = -(cross_val_score(model, X, y, cv=5,scoring="neg_mean_absolute_error"))
    scoreMape = -(cross_val_score(model, X, y, cv=5, scoring="neg_mean_absolute_percentage_error"))
    MEAMean = np.mean(scores) # Calculate the mean MAE across all folds
    MEASTD = np.std(scores) #Calculate the standard deviation MAE across all folds
    MAPEMean = np.mean(scoreMape) # Calculate the mean MAPE across all folds
    MAPESTD = np.std(scoreMape) #Calculate the standard deviation MAPE across all folds
    cross_val_scores.append({"Degree": degree,"rSquareMean":rSqr_mean,"rSquareSTD":rSqr_std,"MEAMean":MEAMean,"MEASTD":MEASTD,"MAPEMean":MAPEMean,"MAPESTD":MAPESTD}) #appending all the metrics of R-square and MAE



c:\Users\20102\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "c:\Users\20102\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 598, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\20102\anaconda3\lib\site-packages\sklearn\pipeline.py", line 341, in fit
    Xt = self._fit(X, y, **fit_params_steps)
  File "c:\Users\20102\anaconda3\lib\site-packages\sklearn\pipeline.py", line 303, in _fit
    X, fitted_transformer = fit_transform_one_cached(
  File "c:\Users\20102\anaconda3\lib\site-packages\joblib\memory.py", line 349, in __call__
    return self.func(*args, **kwargs)
  File "c:\Users\20102\anaconda3\lib\site-packages\sklearn\pipeline.py", line 754, in _fit_transform_one
    res = transformer.fit_transform(X, y, **fit_params)
  File 

In [19]:
cross_val_scores_df = pd.DataFrame(cross_val_scores)
cross_val_scores_df

Degree   rSquareMean     rSquareSTD      MEAMean       MEASTD   MAPEMean  \
0       0           NaN            NaN          NaN          NaN        NaN   
1       1      0.482318       0.049266    44.276402     2.100154   0.394860   
2       2      0.391489       0.120528    46.614216     2.193673   0.402680   
3       3 -90139.144312  179963.620303  1946.414746  3291.215836  16.317068   
4       4    -70.679198      53.259531   303.106752    39.368098   2.453599   
5       5    -67.396319      50.889763   295.589370    37.296266   2.404764   
6       6    -67.455962      51.048883   295.534425    37.322088   2.404479   
7       7    -67.457000      51.051162   295.532935    37.322930   2.404477   
8       8    -67.456859      51.051143   295.532818    37.322773   2.404474   

     MAPESTD  
0        NaN  
1   0.029602  
2   0.029586  
3  28.529223  
4   0.374812  
5   0.352100  
6   0.352515  
7   0.352525  
8   0.352520

In [22]:
Best_Degree_R_Square = cross_val_scores_df["Degree"].iloc[1]
print("BEST Performance for R_Square Metrics among polynomial Degree 0-9  is :",Best_Degree_R_Square)
R_mean = cross_val_scores_df["rSquareMean"].iloc[1]
print("The R_Square metric of Mean is:",round(R_mean,2))
R_STD = cross_val_scores_df["rSquareSTD"].iloc[1]
print("The R_Square metric of Standard-deviation is",round(R_STD,3))
print("\n")
Best_Degree_MAE = cross_val_scores_df["Degree"].iloc[1]
print("BEST Performance for MAE Metrics among polynomial Degree 0-9  is :",Best_Degree_MAE)
MEA_MEAN = cross_val_scores_df["MEAMean"].iloc[1]
print("The MEA metric of Mean is:",round(MEA_MEAN,2))
MEA_STD = cross_val_scores_df["MEASTD"].iloc[1]
print("The MEA metric of Standard-Deviation is:",round(MEA_STD,3))
print("\n")
Best_Degree_MAPE = cross_val_scores_df["Degree"].iloc[1]
print("BEST Performance for MAPE Metrics among polynomial Degree 0-9  is :",Best_Degree_MAPE)
MAPE_MEAN = cross_val_scores_df["MAPEMean"].iloc[1]
print("The MAPE metric of Mean is:",round(MAPE_MEAN,2))
MAPE_STD = cross_val_scores_df["MAPESTD"].iloc[1]
print("The MAPE metric of Standard-Deviation is:",round(MAPE_STD,3))

BEST Performance for R_Square Metrics among polynomial Degree 0-9  is : 1
The R_Square metric of Mean is: 0.48
The R_Square metric of Standard-deviation is 0.049


BEST Performance for MAE Metrics among polynomial Degree 0-9  is : 1
The MEA metric of Mean is: 44.28
The MEA metric of Standard-Deviation is: 2.1


BEST Performance for MAPE Metrics among polynomial Degree 0-9  is : 1
The MAPE metric of Mean is: 0.39
The MAPE metric of Standard-Deviation is: 0.03


### The selection between R-Squared, MAE and MAPE as metrics depends on our specific use case and what we prioritize. In the case of R-Squared, with a value of 0.48, it implies that roughly 48% of the variance in the dependent variable can be accounted for by the independent variables. Moreover, R-Squared typically ranges between 0 and 1, with higher values indicating a superior fit for the regression model, suggesting that our data fits this model better.

### In essence, the decision between R-Squared and MAE hinges on whether we prioritize the goodness of fit (R-Squared) or the accuracy of predictions (MAE).

### MAPE is a percentage error metric, and a lower MAPE indicates a better predictive accuracy. In this case, a MAPE of 0.39 suggests that the model's predictions are, on average, very close to the actual values, with only a 0.39% deviation. It's a relatively small error and indicates a high level of accuracy in the model's predictions.





